src/remoted/shared_download.c

```diff
@@ -9,638 +9,479 @@
-                            }
-                        }
+                os_realloc((*config), sizeof(sd_config_t) * (index + 2), (*config));
+                memset((*config) + index + 1, 0, sizeof(sd_config_t));
+                index++;
+ 
+                if (!strcmp(yaml_node.scalar, "groups")) {
+                    if ((*config)->groups) {
+                        mwarn("Parsing '%s': redefinition of 'group'. Ignoring repeated sections", (*config)->file);
                     } else {
-                        merror("Parsing file '%s': unexpected identifier: '%s' on line %d", yaml_file, w_read_scalar_value(&event), (unsigned int)event.start_mark.line);
+                        if (!sd_parse_groups(&document, yaml_node.value, &(*config)->groups, &(*config)->n_groups)) {
+                            return 0;
+                        }
                     }
-
-                    break;
-                case YAML_MAPPING_END_EVENT:
-                    break;
-
-                default:
-                    merror("Parsing error on line %d: unexpected token", (unsigned int)event.start_mark.line);
-                    goto end;
-                }
-            } while (event.type != YAML_MAPPING_END_EVENT);
-            break;
-
-        default:
-            merror("Parsing error on line %d: unexpected token", (unsigned int)event.start_mark.line);
-            goto end;
+                } else if (!strcmp(yaml_node.scalar, "agents")) {
+                    if (!sd_parse_agents(&document, yaml_node.value, &(*config)->agents, &(*config)->n_agents)) {
+                        return 0;
+                    }
+                } else {
+                    merror("Parsing error on line %d:, unknown token '%s'", (unsigned int)yaml_node.value->start_mark.line, yaml_node.scalar);
+                    return 0;
+                }       
+            } else {
+                merror("Mapping key must be scalar (line %u)", (unsigned int)yaml_node.key->start_mark.line);
+                return 0;
+            }
         }
-
-        break;
-
-    default:
-        mwarn("Parsing '%s': file empty", yaml_file);
-    }
-
-    yaml_event_delete(&event);
-
-    if (!(yaml_parser_parse(&parser, &event) && event.type == YAML_DOCUMENT_END_EVENT)) {
-        merror("Parser error on line %d: [(%d)-(expecting document end)]", (unsigned int)parser.problem_mark.line, parser.error);
-        goto end;
-    }
-
-    yaml_event_delete(&event);
-
-    if (!(yaml_parser_parse(&parser, &event) && event.type == YAML_STREAM_END_EVENT)) {
-        merror("Parser error on line %d: [(%d)-(expecting file end on line)]",(unsigned int)parser.problem_mark.line, parser.error);
-        goto end;
+    } else {
+        merror("Root node must be mapping (line %u)", (unsigned int)root_node->start_mark.line);
+        return 0;
     }
 
-    retval = 1;
-
-end:
-    yaml_event_delete(&event);
     yaml_parser_delete(&parser);
+    yaml_document_delete(&document);
```

Comment:
Looks like there are several cases on which `parser`,  `document`, and `fh` are leaked.